pallets/claims/src/lib.rs

```diff
@@ -0,0 +1,175 @@
+	dispatch::DispatchResult,
+	ensure,
+	traits::Get,
+	weights::{DispatchClass, Pays},
+};
+use frame_system::ensure_signed;
+use hex::FromHex;
+use orml_traits::{MultiCurrency, MultiCurrencyExtended};
+use orml_utilities::with_transaction_result;
+use primitives::{Amount, AssetId, Balance, CORE_ASSET_ID};
+use sp_io::{crypto::secp256k1_ecdsa_recover, hashing::keccak_256};
+use sp_runtime::traits::Zero;
+use sp_std::prelude::*;
+use sp_std::vec::Vec;
+pub use traits::*;
+
+mod claims_data;
+mod traits;
+
+#[cfg(test)]
+mod mock;
+
+#[cfg(test)]
+mod tests;
+
+pub trait Config: frame_system::Config {
+	type Event: From<Event<Self>> + Into<<Self as frame_system::Config>::Event>;
+	type Currency: MultiCurrencyExtended<Self::AccountId, CurrencyId = AssetId, Balance = Balance, Amount = Amount>;
+	type Prefix: Get<&'static [u8]>;
+}
+
+decl_storage! {
+	trait Store for Module<T: Config> as Claims {
+		HDXClaims get(fn hdxclaims): map hasher(blake2_128_concat) EthereumAddress => Balance;
+
+		PalletVersion: StorageVersion = StorageVersion::V1EmptyBalances;
+	}
+
+	add_extra_genesis {
+		config(claims): Vec<(EthereumAddress, Balance)>;
+
+		build(|config: &GenesisConfig| {
+			config.claims.iter().for_each(|(eth_address, initial_balance)| {
+				HDXClaims::mutate(eth_address, | amount | *amount += *initial_balance)
+			})
+		})
+	}
+}
+
+decl_event!(
+	pub enum Event<T>
+	where
+		AccountId = <T as frame_system::Config>::AccountId,
+		Balance = Balance,
+	{
+		HDXClaimed(AccountId, Balance),
+	}
+);
+
+decl_error! {
+	pub enum Error for Module<T: Config> {
+		InvalidEthereumSignature,
+		InvalidStatement,
+		NoClaimOrAlreadyClaimed,
+	}
+}
+
+decl_module! {
+	pub struct Module<T: Config> for enum Call where origin: T::Origin {
+		type Error = Error<T>;
+
+		fn deposit_event() = default;
+
+		/// The Prefix that is used in signed Ethereum messages for this network
+		const Prefix: &[u8] = T::Prefix::get();
+
+		#[weight = (0, DispatchClass::Normal, Pays::No)]
+		fn claim(origin, ethereum_signature: EcdsaSignature)  {
+			let sender = ensure_signed(origin)?;
+
+			let sender_hex = sender.using_encoded(to_ascii_hex);
+
+			let signer = Self::eth_recover(&ethereum_signature, &sender_hex).ok_or(Error::<T>::InvalidEthereumSignature)?;
+
+			Self::process_claim(signer, sender)?;
+		}
+
+		fn on_runtime_upgrade() -> frame_support::weights::Weight {
+			migration::migrate_to_v2::<T>()
+		}
+	}
+}
+
+impl<T: Config> Module<T> {
+	fn process_claim(signer: EthereumAddress, dest: T::AccountId) -> DispatchResult {
+		// TODO: Fix multicurrency support and separate checks for not matching addresses and already claimed
+		let balance_due = HDXClaims::get(&signer);
+
+		ensure!(balance_due != Zero::zero(), Error::<T>::NoClaimOrAlreadyClaimed);
+
+		with_transaction_result(|| {
+			HDXClaims::insert(signer, 0);
+			T::Currency::deposit(CORE_ASSET_ID, &dest, balance_due)?;
+			Ok(())
+		})?;
+
+		Self::deposit_event(RawEvent::HDXClaimed(dest, balance_due));
```

Comment:
actually, you can move the event and OK inside the with transaction and use that to return the result directly.